compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -241,27 +357,52 @@ fn create_generic_args<'tcx>(
     }
 }
 
+// FIXME(fn_delegation): Move generics inheritance to the AST->HIR lowering.
+// For now, generic parameters are not propagated to the generated call,
+// which leads to inference errors:
+//
+// fn foo<T>(x: i32) {}
+//
+// reuse foo as bar;
+// desugaring:
+// fn bar<T>() {
+//   foo::<_>() // ERROR: type annotations needed
+// }
 pub(crate) fn inherit_generics_for_delegation_item<'tcx>(
     tcx: TyCtxt<'tcx>,
     def_id: LocalDefId,
     sig_id: DefId,
-) -> Option<ty::Generics> {
+) -> ty::Generics {
     let builder = GenericsBuilder::new(tcx, sig_id);
+    let parent = tcx.parent(def_id.into());
```

Comment:
The parent is always evaluated anyway, so maybe always set it in the builder, and then use or not use it depending on `InheritanceKind`.

Reply:
It was inlined in the branches, where it was used. `parent` and `InheritanceKind` fields are orthogonal. Generics and predicates are inherited from `sig_id`. `parent` is calculated from `def_id` and then simply propagated to `ty::Generics`, `ty::GenericPredicates`.

Reply:
Ok.